src/AI/WallyAI.java

```diff
@@ -0,0 +1,794 @@
+
+//              log(String.format("    Intermediate waypoint: %s", goal));
+
+              // Sort my currently-reachable move locations by distance from the goal,
+              // and build a GameAction to move to the closest one.
+              Utils.sortLocationsByDistance(goal, destinations);
+              XYCoord destination = null;
+              // try to get somewhere safe
+              log(String.format("    %s would like to travel towards %s safely", unit.toStringWithLocation(), goal));
+              for( XYCoord xyc : destinations )
+              {
+                log(String.format("    is it safe to go to %s?", xyc));
+                if( canWallHere(gameMap, threatMap, unit, xyc) )
+                {
+                  log(String.format("    Yes"));
+                  destination = xyc;
+                  break;
+                }
+              }
+              if( null != destination )
+              {
+                Path movePath = Utils.findShortestPath(unit, destination, gameMap);
+                if( movePath.getPathLength() > 1 ) // We only want to try to travel if we can actually go somewhere
+                {
+                  GameAction move = new GameAction.WaitAction(unit, movePath);
+                  actions.offer(move);
+                  stateChange = true;
+                  break;
+                }
+              }
+            }
+          }
+        }
+      }
+
+      // Check for an available buying enhancement power
+      if( actions.isEmpty() && !stateChange )
+      {
+        AIUtils.queueCromulentAbility(actions, myCo, CommanderAbility.PHASE_BUY);
+      }
+
+      // We will add all build commands at once, since they can't conflict.
+      if( actions.isEmpty() && !stateChange )
+      {
+        queueUnitProductionActions(gameMap);
+      }
+
+      // Check for a turn-ending power
+      if( actions.isEmpty() && !stateChange )
+      {
+        AIUtils.queueCromulentAbility(actions, myCo, CommanderAbility.PHASE_TURN_END);
+      }
+
+      // Return the next action, or null if actions is empty.
+      nextAction = actions.poll();
+    } while (nextAction == null && stateChange); // we don't want to end early, so if the state changed and we don't have an action yet, try again
+    log(String.format("  Action: %s", nextAction));
+    return nextAction;
+  }
+
+  private boolean isSafe(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
+  {
+    Double threat = threatMap.get(myCo.getUnitModel(unit.model.type)).get(xyc);
+    int threshhold = unit.model.hasDirectFireWeapon() ? DIRECT_THREAT_THRESHHOLD : INDIRECT_THREAT_THRESHHOLD;
+    return (null == threat || threshhold > threat);
+  }
+
+  /**
+   * @return whether it's safe or a good place to wall
+   */
+  private boolean canWallHere(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
```

Comment:
This function checks if either 1) `unit` is save (no enemies can substantially threaten it), or 2) `unit` can park next to a more-valuable ally. Right? The current comments ("if we'd be a nice wall for a worthy ally") seems to imply taking into account position relative to the enemy, but that doesn't seem to be the case; we are just as likely to "wall" on the home side of an exposed arty as on the threatened side. Or am I misreading?

Reply:
> This function checks if either 1) unit is safe (no enemies can substantially threaten it), or 2) unit can park next to a more-valuable ally. Right?

Yes.

> The current comments ("if we'd be a nice wall for a worthy ally") seems to imply taking into account position relative to the enemy, but that doesn't seem to be the case; we are just as likely to "wall" on the home side of an exposed arty as on the threatened side. Or am I misreading?

Not quite. This function would have already returned true as the back-side spot *is* actually safe.

I suppose a future version of the function might end up checking if the spot is somewhere an enemy could stand on to attack. Blocking off the enemy such that it can't even approach the spot in the first place would be cool too, but harder to figure out probably.

Reply:
But also, bear in mind that the AI can't currently account for powers at all, so whether an enemy can get here this turn with their current stats doesn't necessarily determine whether walling here is a good idea.